### PR TITLE
Initially implement distributed consistency with jraft

### PR DESCRIPTION
**Pull request by Linary**
Change-Id: I75671444d48a2851b4d1ad0de8975853dd1f900f

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java [hunk 1]
```diff
@@ -343,9 +343,15 @@ private GraphTransaction graphTransaction() {
...
-                 backend, this.name);
-        return BackendProviderFactory.open(backend, this.name);
+        boolean raftMode = this.configuration.get(CoreOptions.RAFT_MODE);
+        if (raftMode) {
+            LOG.info("Opening backend store '{}' in raft mode for graph '{}'",
+                     backend, this.name);
+        } else {
+            LOG.info("Opening backend store '{}' for graph '{}'",
+                     backend, this.name);
+        }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java:353]: can we move the log to BackendProviderFactory.open()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/BackendStoreStateMachine.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        return result;
+    }
+
+    public boolean isLeader() {
+        return this.leaderTerm.get() > 0;
+    }
+
+    @Override
+    public void onLeaderStart(long term) {
+        LOG.debug("Become to Leader");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/BackendStoreStateMachine.java:82]: log who

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStoreProvider.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+/**
+ * Is actually an proxy
+ */
+public class RaftBackendStoreProvider implements BackendStoreProvider {
+
+    private static final Logger LOG = Log.logger(RaftBackendStoreProvider.class);
+
+    // It will be synchronized wrap
+    public static final Map<String, RaftNode> RAFT_NODES = new HashMap<>();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStoreProvider.java:42]: prefer one nodes map each Provider

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+import com.baidu.hugegraph.backend.store.BackendStore;
+import com.baidu.hugegraph.config.CoreOptions;
+import com.baidu.hugegraph.config.HugeConfig;
+import com.baidu.hugegraph.util.Log;
+
+public class RaftNode {
+
+    private static final Logger LOG = Log.logger(RaftNode.class);
+
+    private final StateMachine fsm;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:48]: rename to stateMachine

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        RpcServer rpcServer = component.rpcServer();
+        // 初始化 raft group 服务框架
+        RaftGroupService raftGroupService;
+        raftGroupService = new RaftGroupService(groupId, serverId, nodeOptions,
+                                                rpcServer, true);
+        // 启动
+        return raftGroupService.start(false);
+    }
+
+    public NodeOptions initNodeOptions(HugeConfig config) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:101]: set private

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        } catch (IOException e) {
+            throw new BackendException("Failed to init raft node", e);
+        }
+    }
+
+    public Node node() {
+        return this.node;
+    }
+
+    public Node initRaftNode(BackendStore store, RaftSharedComponent component)
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:64]: set private

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedComponent.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    private final RpcServer rpcServer;
+    private final Executor readIndexExecutor;
+
+    public RaftSharedComponent(HugeConfig config) {
+        this.config = config;
+        this.rpcServer = this.initAndStartRpcServer();
+        this.readIndexExecutor = this.initReadIndexExecutor();
+    }
+
+    public HugeConfig config() {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedComponent.java:50]: set protected

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedComponent.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        return createReadIndexExecutor(Math.max(Utils.cpus() << 2, 4));
+    }
+
+    private static ExecutorService createReadIndexExecutor(final int coreThreads) {
+        final int maxThreads = coreThreads << 2;
+        final RejectedExecutionHandler handler = new ThreadPoolExecutor.AbortPolicy();
+        return newPool(coreThreads, maxThreads, "kvstore-read-index-callback", handler);
+    }
+
+    private static ExecutorService newPool(final int coreThreads, final int maxThreads,
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedComponent.java:79]: delete this method

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedComponent.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                                           final String name,
+                                           final RejectedExecutionHandler handler) {
+        final BlockingQueue<Runnable> defaultWorkQueue = new SynchronousQueue<>();
+        return newPool(coreThreads, maxThreads, defaultWorkQueue, name, handler);
+    }
+
+    private static ExecutorService newPool(final int coreThreads, final int maxThreads,
+                                           final BlockingQueue<Runnable> workQueue,
+                                           final String name,
+                                           final RejectedExecutionHandler handler) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedComponent.java:89]: can remove the final for align

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/BackendStoreStateMachine.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        while (iter.hasNext()) {
+            StoreCommand command;
+            StoreClosure closure = (StoreClosure) iter.done();
+            if (closure != null) {
+                // Leader just take it out from the closure
+                command = closure.command();
+            } else {
+                // Follower need deserializeMutation data
+                ByteBuffer buffer = iter.getData();
+                command = StoreCommand.fromBytes(buffer.array());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/BackendStoreStateMachine.java:57]: pass this.store to StoreCommand

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+        while (iter.hasNext()) {
+            StoreCommand command;
+            StoreClosure closure = (StoreClosure) iter.done();
+            if (closure != null) {
+                // Leader just take it out from the closure
+                command = closure.command();
+            } else {
+                // Follower need deserializeMutation data
+                ByteBuffer buffer = iter.getData();
+                command = StoreCommand.fromBytes(buffer.array());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java:64]: address history comments

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStdSessions.java [hunk 12]
```diff
@@ -592,6 +603,16 @@ public boolean hasChanges() {
             return this.batch.count() > 0;
         }
 
+        @Override
+        public String dataPath() {
+            return dataPath;
```
[the remark this answers is not in the document]
Comment (Linary) [reply hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStdSessions.java:608]: Here is in StdSession class, it's a inner class

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStdSessions.java [hunk 13]
```diff
@@ -592,6 +603,16 @@ public boolean hasChanges() {
...
+        }
+
+        @Override
+        public String walPath() {
+            return walPath;
```
[the remark this answers is not in the document]
Comment (Linary) [reply hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStdSessions.java:613]: Here is in StdSession class, it's a inner class

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedComponent.java (hunk 7 above)
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftSharedComponent.java]: use public keep the same with others

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java (hunk 11 above)
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java:64]: handled

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java [hunk 14]
```diff
@@ -48,50 +49,51 @@
...
             return null;
         });
-        // mutate
-        this.register(StoreCommand.MUTATE, buffer -> {
-            BackendMutation m = StoreSerializer.deserializeMutation(buffer);
-            this.store.mutate(m);
+        this.register(StoreAction.BEGIN_TX, this.store::beginTx);
+        this.register(StoreAction.COMMIT_TX, buffer -> {
+            List<BackendMutation> ms = StoreSerializer.readMutations(buffer);
+            for (BackendMutation mutation : ms) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java:78]: call this.store.beginTx() before mutate()

**Pull request merged**